### PR TITLE
Updated Create Pod+PodStatus resource lifecycle test - +4 endpoint coverage

### LINKED ISSUE #90936: Write CoreV1Pod and CoreV1PodStatus lifecycle test - +4 coverage
# Progress `[7/8]`
* [x] Original approval issue: #88545
* [x]  Test approval issue: [kubernetes/kubernetes#90936](https://github.com/kubernetes/kubernetes/issues/90936)
* [x]  [APISnoop PR](https://github.com/cncf/apisnoop/pull/290)
* [x]  Test pr: [PR](https://github.com/kubernetes/kubernetes/pull/90942) - Reverted due to flakes
* [x] New Test [PR](https://github.com/kubernetes/kubernetes/pull/93459)
* [x]  Two weeks soak start date: 7 November 2020 [testgrid-link](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&include-filter-by-regex=should%20run%20through%20the%20lifecycle%20of%20Pods%20and%20PodStatus) 
* [ ]  Two weeks soak end date: 20 November 2020
* [x] Test promotion pr: [#96485](https://github.com/kubernetes/kubernetes/pull/96485)

This issue is created to allow edit by @Riaankl

/wip
/hold

# Identify an untested feature Using APISnoop
According to this APIsnoop query, there are still some remaining Pod endpoints which are untested.

```
SELECT
  operation_id,
  -- k8s_action,
  -- path,
  -- description,
  kind
  FROM untested_stable_core_endpoints
  where path not like '%volume%'
  and kind like 'Pod'
  -- and kind like ''
  -- and operation_id ilike '%%'
 ORDER BY kind,operation_id desc
 -- LIMIT 25
       ;
```

```
            operation_id             | kind 
-------------------------------------|------
 replaceCoreV1NamespacedPodStatus    | Pod
 readCoreV1NamespacedPodStatus       | Pod
 listCoreV1PodForAllNamespaces       | Pod
 deleteCoreV1CollectionNamespacedPod | Pod
(4 rows)
```

# Use API Reference to Lightly Document the Feature
* [Kubernetes API Reference Docs](https://kubernetes.io/docs/reference/kubernetes-api/)
* [client-go - Pod](https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/core/v1/pod.go)

# The mock test
## Test outline
1. Create a Pod with a static label
2. Patch the Pod with a new Label and updated data
3. Get the Pod to ensure it's patched
4. Read the Pod's status
5. Replace the Pod's status Ready condition to False
6. Read the Pod's status to check if Ready is False
7. List all Pods in all Namespaces find the Pod(1) ensure that the Pod is found and is patched
8. Delete Namespaced Pod(1) via a Collection with a LabelSelector
9. Get the Pod to check that it's deleted

## Example in Go
```go
package main
import (
  "encoding/json"
  "time"
  "fmt"
  "flag"
  "os"
  v1 "k8s.io/api/core/v1"
  "k8s.io/client-go/dynamic"
  "k8s.io/apimachinery/pkg/runtime/schema"
  metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
  "k8s.io/client-go/kubernetes"
  "k8s.io/apimachinery/pkg/types"
  "k8s.io/client-go/tools/clientcmd"
)
func main() {
  // uses the current context in kubeconfig
  kubeconfig := flag.String("kubeconfig", fmt.Sprintf("%v/%v/%v", os.Getenv("HOME"), ".kube", "config"), "(optional) absolute path to the kubeconfig file")
  flag.Parse()
  config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
  if err != nil {
      fmt.Println(err)
      return
  }
  // make our work easier to find in the audit_event queries
  config.UserAgent = "live-test-writing"
  // creates the clientset
  ClientSet, _ := kubernetes.NewForConfig(config)
  DynamicClientSet, _ := dynamic.NewForConfig(config)
  podResource := schema.GroupVersionResource{Group: "", Version: "v1", Resource: "pods"}
  testNs := "default"
  testPodName := "pod-test"
  testPodImage := "nginx"
  testPodImage2 := "nginx"
  testPod := v1.Pod{
      ObjectMeta: metav1.ObjectMeta{
          Name: testPodName,
          Labels: map[string]string{
              "test-pod-static": "true",
          },
      },
      Spec: v1.PodSpec{
          Containers: []v1.Container{
              {
                 Name: testPodName,
                 Image: testPodImage,
              },
          },
      },
  }
  // create Pod with a static label
  _, err = ClientSet.CoreV1().Pods(testNs).Create(&testPod)
  if err != nil {
      fmt.Println(err)
      return
  }
  fmt.Println("[status] created Pod")
  // setup a watch for the RC
  podWatch, err := ClientSet.CoreV1().Pods(testNs).Watch(metav1.ListOptions{LabelSelector: "test-pod-static=true"})
  if err != nil {
      fmt.Println(err)
      return
  }
  podWatchChan := podWatch.ResultChan()
  fmt.Println("[status] watching for Pod to be ready")
  for event := range podWatchChan {
      podEvent, ok := event.Object.(*v1.Pod)
      if ok != true {
         fmt.Println("Unable to fix type")
         return
      }
      if podEvent.Status.Phase == "Running" {
         break
      }
  }
  fmt.Println("[status] Pod is Ready")
  // patch the Pod with a new Label and updated data
  podPatch, err := json.Marshal(map[string]interface{}{
      "metadata": map[string]interface{}{
          "labels": map[string]string{
              "podtemplate": "patched",
          },
      },
      "spec": map[string]interface{}{
          "containers": []map[string]interface{}{
              {
                  "name": testPodName,
                  "image": testPodImage2,
              },
          },
      },
  })
  if err != nil {
      fmt.Println(err)
      return
  }
  _, err = ClientSet.CoreV1().Pods(testNs).Patch(testPodName, types.StrategicMergePatchType, []byte(podPatch))
  if err != nil {
      fmt.Println(err)
      return
  }
  // get the Pod and ensure it's patched
  pod, err := ClientSet.CoreV1().Pods(testNs).Get(testPodName, metav1.GetOptions{})
  if err != nil {
      fmt.Println(err)
      return
  }
  if pod.ObjectMeta.Labels["test-pod-static"] != "true" || pod.Spec.Containers[0].Image != testPodImage2 {
      fmt.Println("[error] patching of Pod failed")
      return
  }
  // get pod status
  podStatusUnstructured, err := DynamicClientSet.Resource(podResource).Namespace(testNs).Get(testPodName, metav1.GetOptions{}, "status")
  if err != nil {
      fmt.Println(err)
      return
  }
  podStatusUjson, _ := json.Marshal(podStatusUnstructured)
  var podStatus v1.Pod
  json.Unmarshal(podStatusUjson, &podStatus)
  // replace the Pod's status Ready condition to False
  podStatusUpdated := podStatus
  podStatusFieldPatchCount := 0
  podStatusFieldPatchCountTotal := 2
  for pos, cond := range podStatusUpdated.Status.Conditions {
      if (cond.Type == "Ready" && cond.Status == "True") || (cond.Type == "ContainersReady" && cond.Status == "True") {
          podStatusUpdated.Status.Conditions[pos].Status = "False"
          podStatusFieldPatchCount++
      }
  }
  if podStatusFieldPatchCount != podStatusFieldPatchCountTotal {
      fmt.Println("[error] failed to patch all relevant Pod conditions")
      return
  }
  _, err = ClientSet.CoreV1().Pods(testNs).UpdateStatus(&podStatusUpdated)
  if err != nil {
      fmt.Println(err)
      return
  }
  fmt.Println("[status] updated PodStatus")
  // list all Pods and get their status to ensure it's Ready condition is False
  podsList, err := ClientSet.CoreV1().Pods("").List(metav1.ListOptions{LabelSelector: "test-pod-static=true"})
  if err != nil {
      fmt.Println(err)
      return
  }
  fmt.Println("[status] fetched all Pods by LabelSelector")
  podStatusFieldPatchCount = 0
  podStatusFieldPatchCountTotal = 2
  for _, podItem := range podsList.Items {
      for _, cond := range podItem.Status.Conditions {
          if (cond.Type == "Ready" && cond.Status == "False") || (cond.Type == "ContainersReady" && cond.Status == "False") {
              podStatusFieldPatchCount++
          }
      }
  }
  if podStatusFieldPatchCount != podStatusFieldPatchCountTotal {
      fmt.Printf("[error] failed to update PodStatus - %v/%v conditions failed to update (%v, %v)", podStatusFieldPatchCount, podStatusFieldPatchCountTotal, "Ready", "ContainersReady")
      return
  }
  fmt.Println("[status] PodStatus was updated successful")
  // delete the Pod via a Collection with a LabelSelector
  err = ClientSet.CoreV1().Pods(testNs).DeleteCollection(&metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: "test-pod-static=true"})
  if err != nil {
      fmt.Println(err)
      return
  }
  fmt.Println("[status] watching for Pod to be not Ready")
  podEventChannel:
  for event := range podWatchChan {
      podEvent, ok := event.Object.(*v1.Pod)
      if ok != true {
          fmt.Println("Unable to fix type")
          return
      }
      podStatusFieldPatchCount := 0
      podStatusFieldPatchCountTotal := 2
      for _, cond := range podEvent.Status.Conditions {
          if (cond.Type == "Ready" && cond.Status == "False") || (cond.Type == "ContainersReady" && cond.Status == "False") {
              podStatusFieldPatchCount++
          }
      }
      if podStatusFieldPatchCount == podStatusFieldPatchCountTotal {
          break podEventChannel
      }
  }
  time.Sleep(5 * time.Second)
  fmt.Println("[status] Pod no longer available")
  // fetch the Pod to check if it's deleted
  _, err = ClientSet.CoreV1().Pods(testNs).Get(testPodName, metav1.GetOptions{})
  if err == nil {
      fmt.Println("[error] Pod still available after deletion; failed to delete Pod")
      return
  }
  // write test here
  fmt.Println("[status] complete")
}
```

```go
[status] created Pod
[status] watching for Pod to be ready
[status] Pod is Ready
[status] updated PodStatus
[status] fetched all Pods by LabelSelector
[status] PodStatus was updated successful
[status] watching for Pod to be not Ready
[status] Pod no longer available
[status] complete
```

# Verify Increase it Coverage with APISnoop
Discover useragents:

```
select distinct useragent from audit_event where bucket='apisnoop' and useragent not like 'kube%' and useragent not like 'coredns%' and useragent not like 'kindnetd%' and useragent like 'live%';
```

```
     useragent     
-------------------
 live-test-writing
(1 row)
```

List endpoints hit by the test:

```
select * from endpoints_hit_by_new_test where useragent like 'live%'; 
```

```
     useragent     |            operation_id             | hit_by_ete | hit_by_new_test 
-------------------|-------------------------------------|------------|-----------------
 live-test-writing | readCoreV1NamespacedPodStatus       | f          |               1
 live-test-writing | readCoreV1NamespacedPod             | t          |               2
 live-test-writing | deleteCoreV1CollectionNamespacedPod | f          |               2
 live-test-writing | patchCoreV1NamespacedPod            | t          |               2
 live-test-writing | listCoreV1PodForAllNamespaces       | f          |               1
 live-test-writing | createCoreV1NamespacedPod           | t          |               2
 live-test-writing | replaceCoreV1NamespacedPodStatus    | f          |               2
 live-test-writing | listCoreV1NamespacedPod             | t          |               1
(8 rows)
```

Display endpoint coverage change:

```
select * from projected_change_in_coverage;
```

```
   category    | total_endpoints | old_coverage | new_coverage | change_in_number 
---------------|-----------------|--------------|--------------|------------------
 test_coverage |             438 |          190 |          194 |                4
(1 row)
```

# Final notes
If a test with these calls gets merged, ****test coverage will go up by 4 points****

This test is also created with the goal of conformance promotion.

/sig testing
/sig architecture
/area conformance

### PR DESCRIPTION
**Pull request by riaankleinhans**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This PR adds a test to test the following untested endpoints:

* replaceCoreV1NamespacedPodStatus
* readCoreV1NamespacedPodStatus
* listCoreV1PodForAllNamespaces
* deleteCoreV1CollectionNamespacedPod

**Which issue(s) this PR fixes**:
Fixes #90936

**Special notes for your reviewer**:
Adds +4 endpoint test coverage (good for conformance)

**Does this PR introduce a user-facing change?**:

```
NONE
```

**Release note:**
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```
NONE
```

/sig testing
/sig architecture
/area conformance

### CODE REVIEW

File: test/e2e/common/pods.go
Comment (dashpole) [inline test/e2e/common/pods.go:998]: Isn't there the possibility that the kubelet has already updated the pod's ready condition back to true here?

Comment (BobyMCbobs) [reply test/e2e/common/pods.go:998]: I might be wrong, but from my understand the response from the API server (in this case `podStatusUpdate`) is basically the same as the request. If that 's true then it should be fine to do this, otherwise I'll remove the checks after the update (so the endpoint is hit)

Comment (dashpole) [reply test/e2e/common/pods.go:998]: Ah, I misread the code, and thought we were doing a separate retrieval

**Review by spiffxp — APPROVED** [review]
/lgtm
/approve
/priority important-soon
/milestone v1.19

**Comment by liggitt** [comment]
/hold

what was modified from the version that was reverted? one of the e2e runs flaked on this test (https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/93459/pull-kubernetes-e2e-gce-ubuntu-containerd/1288972956017364992)

**Comment by spiffxp** [comment]
/lgtm cancel
ack, missed that there had been reverts related to this

**Comment by liggitt** [comment]
At this point, given we know there are latent flakes, I would recommend focusing on root causing and resolving those rather than trying to trigger more failures. Digging into the logs for the existing failures would be where I would start.

**Comment by liggitt** [comment]
I would also recommend moving this to 1.20

The only things going into 1.19 now are release blockers, cleanup things like OWNER file fixup, and things improving CI signal by reducing flakes

**Comment by heyste** [comment]
[PR History](https://prow.k8s.io/pr-history/?org=kubernetes&repo=kubernetes&pr=93459) shows a single clean run. Checking again for any flakes

/test all

**Comment by heyste** [comment]
pull-kubernetes-e2e-kind-ipv6: no such package '@debian-iptables-amd64//image'

/test pull-kubernetes-e2e-kind-ipv6

**Comment by heyste** [comment]
pull-kubernetes-e2e-kind-ipv6: Two failures for e2e suite: [sig-api-machinery] AdmissionWebhook

/test pull-kubernetes-e2e-kind-ipv6

**Comment by spiffxp** [comment]
It's not clear to me that anything has been done here to address @liggitt's concerns.  Spamming /test and /retest is not the same thing as explaining why the flakes listed in https://github.com/kubernetes/kubernetes/pull/93405 won't show up again, or fixing the code to address them

**Comment by MonzElmasry** [comment]
Hello :wave: ! Bug triage here, This PR has not been updated for a long time, so I'd like to check what's the status. The code freeze is starting 12.11.2020 (about 2 weeks from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the PR is tagged for 1.20, is it still planned for this release?

**Comment by Riaankl** [comment]
Hi @MonzElmasry , the plan is to target the PR this week to see if we could get it merge in 1.20

**Comment by heyste** [comment]
Before the new [deflake guide](https://gist.github.com/liggitt/6a3a2217fa5f846b52519acfc0ffece0#deflaking-e2e-tests) was available a lot of time was taken to review the code and check how it worked with a low resource cluster (on a stressed VM) with the goal of checking that it was reliable as possible. I wasn't able to get this test to fail once.

In July there where a lot of test infrastructure issues that created a lot of noise that may of masked an unknown issue with the test. The deflake guide has some great advice that will be used for any future issues.

The goal is to get an updated set of logs that show that this PR is solid before asking for a merge request.

**Comment by heyste** [comment]
/test pull-kubernetes-e2e-gce-ubuntu-containerd
flake
```
Kubernetes  e2e suite: [sig-storage] PersistentVolumes NFS with Single PV - PVC  pairs create a PV and a pre-bound PVC: test write access

pod "pvc-tester-nbnvp" did not exit with Success: pod "pvc-tester-nbnvp" failed to reach Success: Gave up after waiting 5m0s for pod "pvc-tester-nbnvp" to be "Succeeded or Failed"
```

**Comment by heyste** [comment]
pull-kubernetes-e2e-kind

```
test/e2e/common/pods.go:886
Oct 29 15:50:50.851: failed to see DELETED event
Unexpected error:
    <*errors.errorString | 0xc00020a200>: {
        s: "timed out waiting for the condition",
    }
    timed out waiting for the condition
occurred
test/e2e/common/pods.go:1031
```
Added a counter to the set of combined logs and then did a rough cut for this test. Test starts 196914 with the failure at 217396
[PR93459-pull-kubernetes-e2e-kind-1321838021317758976-test-fail.log](https://github.com/kubernetes/kubernetes/files/5461266/PR93459-pull-kubernetes-e2e-kind-1321838021317758976-test-fail.log)

**Comment by heyste** [comment]
/test pull-kubernetes-e2e-gce
flake

```
Kubernetes  e2e suite: [sig-cli] Kubectl client Kubectl server-side dry-run should  check if kubectl can dry-run update Pods [Conformance]

test/e2e/framework/framework.go:629
Oct 30 18:31:21.234: Unexpected error:
...
Error from server (Conflict): error when replacing \"STDIN\": Operation cannot be fulfilled on pods \"e2e-test-httpd-pod\": the object has been modified; please apply your changes to the latest version and try again\n\nerror:\nexit status 1"
```

**Comment by heyste** [comment]
/test pull-kubernetes-e2e-gce-ubuntu-containerd
flake

```
Kubernetes  e2e suite: [sig-cli] Kubectl client Kubectl server-side dry-run should  check if kubectl can dry-run update Pods [Conformance]

test/e2e/framework/framework.go:629
Oct 30 18:09:28.719: Unexpected error:
...
Error from server (Conflict): error when replacing \"STDIN\": Operation cannot be fulfilled on pods \"e2e-test-httpd-pod\": the object has been modified; please apply your changes to the latest version and try again\n\nerror:\nexit status 1"
```

**Comment by heyste** [comment]
/test pull-kubernetes-e2e-gce
flake

`Kubernetes e2e suite: [sig-cli] Kubectl client Kubectl server-side dry-run should check if kubectl can dry-run update Pods [Conformance]` 
(same error message as listed before)

**Comment by heyste** [comment]
/test pull-kubernetes-e2e-gce-ubuntu-containerd
flake
```
Kubernetes e2e suite: [sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: dir-link-bindmounted] [Testpattern: Pre-provisioned PV (default fs)] subPath should support existing single file [LinuxOnly]
test/e2e/storage/testsuites/subpath.go:216
Nov  3 01:56:44.281: Unexpected error:
    <*errors.errorString | 0xc002891f70>: {
...
Terminated:&ContainerStateTerminated{ExitCode:128,Signal:0,Reason:StartError,Message:failed to create containerd task: failed to start io pipe copy: unable to copy pipes: containerd-shim: opening w/o fifo \"/run/containerd/io.containerd.grpc.v1.cri/containers/8bb932f8db7b7d6429405e1f90c08c51a2d1f02b001b9c7d784ccc76861e18cd/io/055386896/8bb932f8db7b7d6429405e1f90c08c51a2d1f02b001b9c7d784ccc76861e18cd-stdout\" failed: context deadline exceeded,
```

**Comment by hh** [comment]
/hold 
For @smarterclayton

**Comment by smarterclayton** [comment]
I don't see a resolution of the question https://github.com/kubernetes/kubernetes/pull/93459#issuecomment-712453789 from @spiffxp, please address.

**Comment by Riaankl** [comment]
@smarterclayton @spiffxp 

Before the new [deflake guide](https://gist.github.com/liggitt/6a3a2217fa5f846b52519acfc0ffece0) was available a lot of time was taken to review the code and check how it worked with a low resource cluster (on a stressed VM) with the goal of checking that it was reliable as possible. I wasn't able to get this test to fail once.

In July there where a lot of test infrastructure issues that created a lot of noise that may of masked an unknown issue with the test. The deflake guide has some great advice that will be used for any future issues.

The goal was to get an updated set of logs that show that this PR is solid before asking for a merge request.
Over the last 10 day's @heyste and I run the proposed test repeatedly to try and expose any issues with the test before merging. 

From all the runs we believe the test ready to go on the test grid.

**Comment by heyste** [comment]
There has been an [update to the test](https://github.com/kubernetes/kubernetes/pull/93459/commits/5ad32fcd741acb4bd5b7f084a6406bb502ad677d) that I believe now fixes a possible flake when the cluster is under stress. Since that update the test hasn't flaked.

File: test/e2e/common/pods.go
Comment (spiffxp) [inline test/e2e/common/pods.go:922]: please refactor this into a named constant or var

**Review by spiffxp — CHANGES_REQUESTED** [review]
I'm not a fan of arbitrarily raising timeouts, but I see the data is proving this out.  What is the source of this number?  Is there some other component timeout that we're trying to out-wait?

File: test/e2e/common/pods.go
Comment (heyste) [reply test/e2e/common/pods.go:922]: It was the timeout value I had used in [e2e test: should delete a collection of pod templates](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/common/podtemplates.go#L38). The latest flake had multiple "throttling" log messages but couldn't see a single reason for it.

**Comment by heyste** [comment]
/test pull-kubernetes-e2e-gce-ubuntu-containerd
flake

`[sig-auth] ServiceAccounts should guarantee kube-root-ca.crt exist in any namespace [It]`
`{"msg":"FAILED [sig-auth] ServiceAccounts should guarantee kube-root-ca.crt exist in any namespace","total":-1,"completed":19,"skipped":141,"failed":1,"failures":["[sig-auth] ServiceAccounts should guarantee kube-root-ca.crt exist in any namespace"]}`

**Comment by liggitt** [comment]
fix for that flake in https://github.com/kubernetes/kubernetes/pull/96274

**Comment by heyste** [comment]
/test pull-kubernetes-e2e-gce-ubuntu-containerd
flakes
```
Kubernetes e2e suite: [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] subPath should support readOnly file specified in the volumeMount [LinuxOnly]
...
expected pod "pod-subpath-test-dynamicpv-r77v" success: Gave up after waiting 5m0s for pod "pod-subpath-test-dynamicpv-r77v" to be "Succeeded or Failed"
```

```
Kubernetes e2e suite: [sig-storage] In-tree Volumes [Driver: gluster] [Testpattern: Pre-provisioned PV (filesystem volmode)] volumeMode should not mount / map unused volumes in a pod
...
error checking directory /var/lib/kubelet/pods/ed5b351a-b2c6-44b4-b894-5a17ea54b5c4/volumeDevices on node e2e-c10f5eb34f-a7d53-minion-group-c0lk: error dialing backend: dial tcp 10.40.0.4:10250: connect: connection refused
```

```
Kubernetes e2e suite: [sig-network] Networking Granular Checks: Services should function for pod-Service: udp
...
Nov  5 23:09:46.105: failed dialing endpoint, did not find expected responses... 
Tries 39
Command curl -g -q -s 'http://10.64.1.234:8080/dial?request=hostname&protocol=udp&host=10.0.252.8&port=90&tries=1'
retrieved map[netserver-0:{} netserver-2:{}]
expected map[netserver-0:{} netserver-1:{} netserver-2:{}]
vendor/github.com/onsi/ginkgo/internal/leafnodes/runner.go:113
```

**Comment by heyste** [comment]
@spiffxp Added a constant as suggested, [PodReadyTimeout](https://github.com/kubernetes/kubernetes/commit/d8099c08d867a91b3cd6dcb3c9d32b10cd788179) and checks are looking good now.

**Comment by spiffxp** [comment]
/approve
needs rebase, ping me for lgtm after that

**Comment by hh** [comment]
Rebased

**Review by spiffxp — COMMENTED** [review]
/lgtm

**Comment by hh** [comment]
To give a quick summary of those last two failures:

9 unrelated tests seem to have connection and node failures:
- https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/93459/pull-kubernetes-e2e-gce-ubuntu-containerd/1324868071680118784/
```
Error from server: Get "https://10.40.0.5:10250/containerLogs/gcp-volume-9890/nfs-server/nfs-server":
dial tcp 10.40.0.5:10250: connect: connection refused
Nov  7 00:43:58.066: failed dialing endpoint (initial), did not find expected responses... 
Message: "error dialing backend: dial tcp 10.40.0.5:10250: i/o timeout",
Message: "an error on the server (\"unknown\") has prevented the request from succeeding (get nodes
```

Device Manager related:
- https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/93459/pull-kubernetes-bazel-test/1324868071638175744#1:build-log.txt%3A164
```
FAIL: //pkg/kubelet/cm/devicemanager:go_default_test (run 1 of 2)...
INFO: From Testing //pkg/kubelet/cm/devicemanager:go_default_test (run 1 of 2):
I1107 00:43:46.703222  215173 device_plugin_stub.go:142]
  Starting to serve on /tmp/device_plugin095116164/device-plugin.sock.third
--- FAIL: TestDevicePluginReRegistrationProbeMode (1.03s)
```

**Pull request merged**
